### ISSUE TITLE
Decoder dropout

### ISSUE THREAD

**Issue by Lguyogiro** [machamp-nlp/machamp#29]
I've been experimenting (mostly successfully) with adding dropout in various decoders, and was wondering if it might be useful to add this option system-wide? For example, enable a "dropout" parameter for any of the given decoders. This is straightforward for e.g. seq and bio_seq tasks, but I'll have to look more in detail at the implementation of the other tasks to see how it would be executed.

Is this something that would be valuable and/or of interest to the package developers? If so, I'm happy to open a PR with, for instance, the decoder dropout implementation for the sequence tasks.

**Comment by robvanderg** [comment]
Interesting, thanks for the idea!, in the very first version we had this, but back then performance differences (on the test we tested for) where negligible, so we dropped it to make the hyperparameter space smaller and the code more compact. If you managed to get performance gains, I'd say it would make sense to re-add it again. 

That being said, I think it would be nicest if its available for (almost) all decoders. I just checked all the decoders, and just searching for "torch." actually reveals all parameters that this should be applied to for as far as I can see. I think the dropout itself can then be saved in machamp_decoder (to avoid redundancy). The only problematic case is MLM, as there we make use of the existing implementation in Huggingface, it could throw a warning if these are used together?

**Comment by Lguyogiro** [comment]
Okay, I'll try to get a PR soon and see what you think. I can also try to test an a few additional datasets/tasks to confirm the gains are worth adding the functionality. In general, it seems like potentially a good option to have for experimentation; in my case a model was/is overfitting on one among multiple tasks, and adding dropout seems to have provided some improvement.

**Comment by robvanderg** [comment]
Great! Additional experiments are only necessary if you want the default to be different from 0.0 I think.

**Comment by Lguyogiro** [comment]
Sorry, I kind of forgot about this. There is an implementation of decoder dropout [here](https://github.com/Lguyogiro/machamp/tree/decoder-dropout). My previous findings that it consistently improved performance for some tasks turned out to be complicated by other factors, so I'm no longer confident this is as valuable as I previously thought. It changes the model behavior, but I no longer found that adding decoder drop out gave me the best results compared to other changes.

If there is any desire to have it, I am happy to merge the newer changes and open a PR. Otherwise we can just kill this issue and that branch.

**Comment by robvanderg** [comment]
Thanks for the update!, As long as the default is set to 0.0 (off), and it then doesn't affect performance, I would be fine with merging it in.
